### PR TITLE
Crosshair snapping

### PR DESCRIPTION
**Pull request by EniacMlezi**
Created a way for making the crosshair "snap" to the closest available ChartPoint. Can be enabled/disabled on a per axis basis

https://user-images.githubusercontent.com/15785664/196917984-4d97c5dc-eab4-40e6-be70-b40f24f90f07.mp4

### CODE REVIEW

File: src/LiveChartsCore/Axis.cs [hunk 1]
```diff
@@ -611,13 +615,43 @@ public void InvalidateCrosshair(Chart<TDrawingContext> chart, LvcPoint pointerPo
...
-            x = pointerPosition.X;
+            float crosshairX;
+            if (CrosshairSnapEnabled)
+            {
+                var axisIndex = Array.IndexOf(cartesianChart.XAxes, this);
+                if (axisIndex == -1) return;
+
+                var closestPoint = FindClosestPoint(pointerPosition, cartesianChart, cartesianChart.Series.Where(s => s.ScalesXAt == axisIndex));
+
+                crosshairX = scale.ToPixels(closestPoint?.SecondaryValue ?? pointerPosition.X);
```
Comment (EniacMlezi) [inline src/LiveChartsCore/Axis.cs:626]: If no close point found, reverts to non-snapping crosshair. Do we want this implementation?

File: src/LiveChartsCore/Axis.cs [hunk 2]
```diff
@@ -686,6 +720,25 @@ public void InvalidateCrosshair(Chart<TDrawingContext> chart, LvcPoint pointerPo
         chart.Update();
     }
 
+    private static ChartPoint? FindClosestPoint(LvcPoint pointerPosition, CartesianChart<TDrawingContext> cartesianChart, IEnumerable<ICartesianSeries<TDrawingContext>> allSeries)
+    {
+        ChartPoint? closestPoint = null;
+        foreach (var series in allSeries)
+        {
+            var hitpoints = series.FindHitPoints(cartesianChart, pointerPosition, TooltipFindingStrategy.CompareOnlyXTakeClosest);
```
Comment (EniacMlezi) [inline src/LiveChartsCore/Axis.cs]: I've tried different "CompareAllTakeClosest", but that gives weird behavior I don't completely understand

File: src/LiveChartsCore/Axis.cs [hunk 3]
```diff
@@ -686,6 +720,25 @@ public void InvalidateCrosshair(Chart<TDrawingContext> chart, LvcPoint pointerPo
...
+            var hitpoint = hitpoints.SingleOrDefault();
+            if (hitpoint == null) continue;
```
Comment (EniacMlezi) [inline src/LiveChartsCore/Axis.cs:732]: No idea how "TakeClosest" can return no elements, but it happens sometimes. Therefore implement a continue here

File: src/LiveChartsCore/Axis.cs [hunk 4]
```diff
@@ -611,13 +616,47 @@ public void InvalidateCrosshair(Chart<TDrawingContext> chart, LvcPoint pointerPo
         float x, y;
         if (_orientation == AxisOrientation.X)
         {
-            x = pointerPosition.X;
+            float crosshairX;
+            if (CrosshairSnapEnabled)
+            {
+                var axisIndex = Array.IndexOf(cartesianChart.XAxes, this);
+                if (axisIndex == -1) return;
```
Comment (EniacMlezi) [inline src/LiveChartsCore/Axis.cs]: return. or try drawing anyways?

File: src/LiveChartsCore/Axis.cs [hunk 5]
```diff
@@ -611,13 +616,47 @@ public void InvalidateCrosshair(Chart<TDrawingContext> chart, LvcPoint pointerPo
...
+            x = crosshairX;
             y = yoo;
         }
         else
         {
+            float crosshairY;
+            if (CrosshairSnapEnabled)
+            {
+                var axisIndex = Array.IndexOf(cartesianChart.YAxes, this);
+                if (axisIndex == -1) return;
```
Comment (EniacMlezi) [inline src/LiveChartsCore/Axis.cs]: return, or try drawing anyways?

**Comment by beto-rodriguez** [comment]
This looks amazing! thanks I will review it and let you know any observation.

**Comment by EniacMlezi** [comment]
> This looks amazing! thanks I will review it and let you know any observation.

Yes, please do! I used FindHitPoints to get the closest ChartPoint, but I bet those calculations have already been done by the ToolTip finding logic. Maybe this functionality can be hooked into that.

File: src/LiveChartsCore/Axis.cs (hunk 1 above)
Comment (beto-rodriguez) [reply src/LiveChartsCore/Axis.cs:626]: Maybe this is a good solution, one case would be when there are no points in the chart, I think it is a good default to follow the cursor while there is no data in the chart yet.

File: src/LiveChartsCore/Axis.cs (hunk 2 above)
Comment (beto-rodriguez) [reply src/LiveChartsCore/Axis.cs]: Just let the library decide that for you, it considers more factors depending on the user settings https://github.com/beto-rodriguez/LiveCharts2/pull/701/commits/d411224a4c3424b2fb56753a1d129ae2e9942650

File: src/LiveChartsCore/Axis.cs (hunk 3 above)
Comment (beto-rodriguez) [reply src/LiveChartsCore/Axis.cs:732]: Stange that take closes has no elements, but besides that it is better to use `FirstOrDefault` in this case:

> Single will throw an exception if it finds more than one record matching the criteria. First will always select the first record from the list

We can have ties, so for this library we need `First` instead.

File: src/LiveChartsCore/Axis.cs [hunk 6]
```diff
@@ -611,13 +616,45 @@ public void InvalidateCrosshair(Chart<TDrawingContext> chart, LvcPoint pointerPo
         float x, y;
         if (_orientation == AxisOrientation.X)
         {
-            x = pointerPosition.X;
+            float crosshairX;
+            if (CrosshairSnapEnabled)
+            {
+                var axisIndex = Array.IndexOf(cartesianChart.XAxes, this);
```
Comment (beto-rodriguez) [inline src/LiveChartsCore/Axis.cs:622]: This is the only thing I don't completely like, we do not need to calculate the index, instead I think that the library should expose a read only property where we can get this value, The library internally can assign this index, there is no need to go and find it every time the crosshair is invalidated.

We can build that in a new PR, for now this seems ok.

**Pull request merged**